llama.cpp

```diff
@@ -1297,6 +1351,9 @@ struct llama_model {
 
     int n_gpu_layers;
 
+    // gguf metadata
+    std::unordered_map<std::string, std::string> gguf_kv;
```

Comment:
What's this doing here then? Why aren't you taking your own suggestion to just `gguf_init_from_file` again?

All I'm saying is save this information in a format that's not only useful for Jinja templates or whatever. Where there's a way to reliably access it and something like `1` can be differentiated from `"1"`.

Reply:
I am not trying to be snarky, but it would take a lot more code to re-load the gguf when using the metadata loading functions, and the only purpose would be to save a few bytes of memory (the reason arrays are excluded is precisely because it would waste too much memory for something of questionable utility). So doing that would be another case of premature optimization.

To be clear, I am not opposed to extending the API to support more types of metadata, that's why the function are called `llama_model_meta_val_str` and not just `llama_model_meta_val`, I expected that somebody may want to extend the API in the future. I am just not convinced that it is really worth the effort and complexity to do that because the expected use of these functions is to get very specific metadata like the jinja templates, which is trivial to convert to a number if needed, or just to dump a bunch of useless info to the user, in which case it doesn't matter what the type is. Making everything accesible as a string is simpler than having to check the correct function to call depending on the type. If this was a C++17 library I would be happy to make it return a `std::variant`, but that's not what it is.